common/quotas/clocked_rate_limiter.go

```diff
@@ -0,0 +1,154 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package quotas
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"golang.org/x/time/rate"
+)
+
+// ClockedRateLimiter wraps a rate.Limiter with a clockwork.Clock. It is used to ensure that the rate limiter respects
+// the time determined by the clock.
+type ClockedRateLimiter struct {
+	rateLimiter *rate.Limiter
+	clock       clockwork.Clock
+}
+
+var (
+	ErrRateLimiterWaitInterrupted                       = errors.New("rate limiter wait interrupted")
+	ErrRateLimiterReservationCannotBeMade               = errors.New("rate limiter reservation cannot be made due to insufficient quota")
+	ErrRateLimiterReservationWouldExceedContextDeadline = errors.New("rate limiter reservation would exceed context deadline")
+)
+
+func NewClockedRateLimiter(rateLimiter *rate.Limiter, clock clockwork.Clock) ClockedRateLimiter {
+	return ClockedRateLimiter{
+		rateLimiter: rateLimiter,
+		clock:       clock,
+	}
+}
+
+func (l ClockedRateLimiter) Allow() bool {
+	return l.AllowN(l.clock.Now(), 1)
+}
+
+func (l ClockedRateLimiter) AllowN(now time.Time, token int) bool {
+	return l.rateLimiter.AllowN(now, token)
+}
+
+// ClockedReservation wraps a rate.Reservation with a clockwork.Clock. It is used to ensure that the reservation
+// respects the time determined by the clock.
+type ClockedReservation struct {
+	reservation *rate.Reservation
+	clock       clockwork.Clock
+}
+
+func (r ClockedReservation) OK() bool {
+	return r.reservation.OK()
+}
+
+func (r ClockedReservation) Delay() time.Duration {
+	return r.DelayFrom(r.clock.Now())
+}
+
+func (r ClockedReservation) DelayFrom(t time.Time) time.Duration {
+	return r.reservation.DelayFrom(t)
+}
+
+func (r ClockedReservation) Cancel() {
+	r.CancelAt(r.clock.Now())
+}
+
+func (r ClockedReservation) CancelAt(t time.Time) {
+	r.reservation.CancelAt(t)
+}
+
+func (l ClockedRateLimiter) Reserve() ClockedReservation {
+	return l.ReserveN(l.clock.Now(), 1)
+}
+
+func (l ClockedRateLimiter) ReserveN(now time.Time, token int) ClockedReservation {
+	reservation := l.rateLimiter.ReserveN(now, token)
+	return ClockedReservation{reservation, l.clock}
+}
+
+func (l ClockedRateLimiter) Wait(ctx context.Context) error {
+	return l.WaitN(ctx, 1)
+}
+
+// WaitN is the only method that is different from rate.Limiter. We need to fully reimplement this method because
+// the original method uses time.Now(), and does not allow us to pass in a time.Time. Fortunately, it can be built on
+// top of ReserveN. However, there are some optimizations that we can make.
+func (l ClockedRateLimiter) WaitN(ctx context.Context, token int) error {
+	reservation := ClockedReservation{l.rateLimiter.ReserveN(l.clock.Now(), token), l.clock}
```

Comment:
looks like this will do reservation & cancellation more often than the original impl (when token can't be get within the wait duration), so more likely run into https://github.com/golang/go/issues/56924 this issue: keep doing reservation & cancellation result in effectively a lower rps limit (we saw that in prod before)

But I think the change is ok since the only way to avoid ^ issue is to only call Wait() with context that has no deadline, and it still holds for with this change. so I guess this comment is just FYI.

Reply:
Tbh I don't fully understand this issue, but it looks like the Go maintainers closed it because it's working as intended?